test/rules/no-unnecessary-type-assertion/test.ts.fix

```diff
@@ -83,3 +83,10 @@ const data = {
 
 [1, 2, 3, 4, 5].map(x => [x, 'A' + x] as [number, string]);
 let x: Array<[number, string]> = [1, 2, 3, 4, 5].map(x => [x, 'A' + x] as [number, string]);
+
+function foo<T>(): T {}
+function bar(s: string) {}
+
+// Type error without this assertion
+const xyzzy = foo() as string;
```

Comment:
If this is the only false positive, I don't think it's worth sacrificing the normal case

Reply:
It's come up several times for us. It's not a common pattern, but it does happen with e.g. Angular's .d.ts

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/angular/index.d.ts#L2058

Reply:
Instead of asserting the type you can just explicitly declare the type argument on the function call: 

```ts
const v = foo<string>();
```

Reply:
That's true, and that's what I suggest when people file bugs. But people keep filing bugs, and I am a limited resource.

Reply:
You could just change the error message to mention that there are cases like this, and recommend using a contextual type as in `const x: string = foo()` or  an explicit type argument as in `const x = foo<string>();` in these cases. The type assertion is still in some sense unnecessary as the code can be written without any type assertions, which is good since those are unsafe. `foo() as string` should really be reserved for functions like `declare function foo(): string | number;`.